### PR TITLE
refactor(ingest): clean up pipeline init error handling

### PR DESCRIPTION
**Pull request by hsheth2**
This is a PoC - I'm hoping we can use a similar pattern in `Report` to call `report_failure` or `report_warning`, which should help us reduce the `except Exception` sprawl and improve error reporting.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/run/pipeline.py [hunk 1]
```diff
@@ -101,6 +106,16 @@ class PipelineInitError(Exception):
...
 
 
+@contextlib.contextmanager
+def _add_init_error_context(step: str) -> Iterator[None]:
+    """Enriches any exceptions raised with information about the step that failed."""
+
+    try:
+        yield
+    except Exception as e:
+        raise PipelineInitError(f"Failed to {step}: {e}") from e
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/run/pipeline.py:116]: Failed while executing step {step}?

File: metadata-ingestion/src/datahub/ingestion/run/pipeline.py [hunk 2]
```diff
@@ -146,7 +158,7 @@ def __init__(
         self.last_time_printed = int(time.time())
         self.cli_report = CliReport()
 
-        try:
+        with _add_init_error_context("set up framework context"):
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/run/pipeline.py:161]: I really like this pattern..

File: metadata-ingestion/src/datahub/ingestion/run/pipeline.py [hunk 3]
```diff
@@ -512,7 +503,7 @@ def pretty_print_summary(
             )
             num_failures_sink = len(self.sink.get_report().failures)
             click.secho(
-                f"{'⏳' if currently_running else ''} Pipeline {'running' if currently_running else 'finished'} with at least {num_failures_source+num_failures_sink} failures {'so far' if currently_running else ''}; produced {workunits_produced} events {duration_message}",
+                f"{'⏳' if currently_running else ''} Pipeline {'running' if currently_running else 'finished'} with at least {num_failures_source+num_failures_sink} failures{' so far' if currently_running else ''}; produced {workunits_produced} events {duration_message}",
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/run/pipeline.py:506]: is this change intentional? cannot tell

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/run/pipeline.py:506]: yup it was intentional - it removes the space from the print of `failures ;`

File: metadata-ingestion/src/datahub/ingestion/run/pipeline.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/run/pipeline.py:116]: This keeps the error messages the same as before e.g. "Failed to set up framework context" - i didn't want to make the errors too verbose

**Review by jjoyce0510 — APPROVED** [review]

**Pull request merged**